### PR TITLE
planner: introduce some new hints for MPP plans

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #xxx

Problem Summary: planner: introduce some new hints for MPP plans

### What is changed and how it works?

Introduce some new hints for MPP plans: mpp_1phase_agg(), mpp_2phase_agg(), mpp_tidb_agg(), mpp_scalar_agg(), shuffle_join().

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/exhaust_physical_plans.go [hunk 1]
```diff
@@ -1940,16 +1940,18 @@ func (p *LogicalJoin) exhaustPhysicalPlans(prop *property.PhysicalProperty) ([]P
...
-				return mppJoins, true, nil
-			}
-			joins = append(joins, mppJoins...)
-		} else {
-			mppJoins := p.tryToGetMppHashJoin(prop, false)
-			joins = append(joins, mppJoins...)
+		shuffleJoins := p.tryToGetMppHashJoin(prop, false)
+		if (p.preferJoinType&preferShuffleJoin) > 0 && len(shuffleJoins) > 0 { // has shuffle_join hints
+			return shuffleJoins, true, nil
+		}
```
Comment (Reminiscent) [inline planner/core/exhaust_physical_plans.go:1946]: If we have both BCJoin hint and shuffle join hint, it seems we will chose the shuffle join. Is it expected?

File: planner/core/exhaust_physical_plans.go [hunk 2]
```diff
@@ -2803,6 +2805,31 @@ func (la *LogicalAggregation) tryToGetMppHashAggs(prop *property.PhysicalPropert
...
+	var preferModes []AggMppRunMode
+	if la.aggHints.preferAggType&preferMPP1PhaseAgg > 0 {
+		preferModes = append(preferModes, Mpp1Phase)
+	} else if la.aggHints.preferAggType&preferMPP2PhaseAgg > 0 {
+		preferModes = append(preferModes, Mpp2Phase)
+	} else if la.aggHints.preferAggType&preferMPPTiDBAgg > 0 {
+		preferModes = append(preferModes, MppTiDB)
+	} else if la.aggHints.preferAggType&preferMPPScalarAgg > 0 {
+		preferModes = append(preferModes, MppScalar)
+	}
```
Comment (Reminiscent) [inline planner/core/exhaust_physical_plans.go:2819]: If there are more one mpp aggregation hints. We can only use one?
Change this part to
```
if .. {}
if .. {}
if .. {}
if .. {}
```

File: planner/core/exhaust_physical_plans.go [hunk 3]
```diff
@@ -2803,6 +2805,31 @@ func (la *LogicalAggregation) tryToGetMppHashAggs(prop *property.PhysicalPropert
...
+	if len(preferModes) > 0 {
+		var preferPlans []PhysicalPlan
+		for _, agg := range hashAggs {
+			hg := agg.(*PhysicalHashAgg)
+			for _, mode := range preferModes {
+				if hg.MppRunMode == mode {
+					preferPlans = append(preferPlans, hg)
+					break
```
Comment (Reminiscent) [inline planner/core/exhaust_physical_plans.go:2827]: Ditto. We should handle the situation which there are many hints.

**Review by Reminiscent — APPROVED** [review]
Rest LGTM

File: planner/core/exhaust_physical_plans.go (hunk 1 above)
Comment (qw4990) [reply planner/core/exhaust_physical_plans.go]: The behavior of using multiple conflict hints together seems undefined. Let's notify our users on docs and any expectation wouldn't be guaranteed here.

**Review by xuyifangreeneyes — APPROVED** [review]

File: planner/core/exhaust_physical_plans.go [hunk 4]
```diff
@@ -1940,16 +1940,18 @@ func (p *LogicalJoin) exhaustPhysicalPlans(prop *property.PhysicalProperty) ([]P
...
+			return shuffleJoins, true, nil
 		}
+		bcastJoins := p.tryToGetMppHashJoin(prop, true)
+		if len(bcastJoins) > 0 &&
+			((p.preferJoinType&preferBCJoin) > 0 || // specified by bcast_join hints
+				p.shouldUseMPPBCJ()) { // specified by heuristic rules
+			return bcastJoins, true, nil
+		}
+		joins = append(joins, bcastJoins...)
+		joins = append(joins, shuffleJoins...)
```
Comment (xuyifangreeneyes) [inline planner/core/exhaust_physical_plans.go:1954]: If there is no hint and `p.shouldUseMPPBCJ` is false, the old code only adds `shuffleJoins` but the new code adds both `bcastJoins` and `shuffleJoins`. Is that an expected change?

**Pull request merged**